idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceAssociateFunctionInspection.kt

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.inspections
+
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.codeInspection.ProblemHighlightType.GENERIC_ERROR_OR_WARNING
+import com.intellij.codeInspection.ProblemHighlightType.INFORMATION
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.openapi.project.Project
+import org.jetbrains.kotlin.builtins.KotlinBuiltIns
+import org.jetbrains.kotlin.config.LanguageVersion
+import org.jetbrains.kotlin.descriptors.ValueParameterDescriptor
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.idea.core.getLastLambdaExpression
+import org.jetbrains.kotlin.idea.inspections.AssociateFunction.*
+import org.jetbrains.kotlin.idea.inspections.collections.isCalling
+import org.jetbrains.kotlin.idea.intentions.callExpression
+import org.jetbrains.kotlin.idea.project.languageVersionSettings
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.anyDescendantOfType
+import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+
+class ReplaceAssociateFunctionInspection : AbstractKotlinInspection() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) = dotQualifiedExpressionVisitor(fun(dotQualifiedExpression) {
+        if (dotQualifiedExpression.languageVersionSettings.languageVersion < LanguageVersion.KOTLIN_1_3) return
+        val callExpression = dotQualifiedExpression.callExpression ?: return
+        val calleeExpression = callExpression.calleeExpression ?: return
+
+        val context = dotQualifiedExpression.analyze(BodyResolveMode.PARTIAL)
+        val isAssociate = callExpression.isCalling(FqName("kotlin.collections.associate"), context)
```

Comment:
I have two concerns here: 

- I would extract all four FqNames into companion object properties. 
- We call `analyze` on each dot qualified expression, which is slow (yes, it's slow in practice, you can try to launch this inspection on Kotlin project). `isCalling` could solve this problem, but unfortunately you call `analyze` before. It's also bad solution to remove `context` here. I think we should first check that callee name starts with `associate` and then check that function is resolved to one of four (better not via `isCalling` because it would repeat the same text check).